pkg/build/golang/docker.go

```diff
@@ -174,30 +135,36 @@ func (b *DockerGoBuilder) Build(ctx context.Context, in *api.BuildInput, ow *rpc
 	// If we have version overrides, apply them.
 	var replaces []string
 	for mod, ver := range in.Dependencies {
-		// TODO(RK): allow to override target of replaces, so we can test against forks.
 		replaces = append(replaces, fmt.Sprintf("-replace=%s=%s@%s", mod, mod, ver))
 	}
 
 	// Inject replace directives for the SDK modules.
-	replaces = append(replaces,
-		"-replace=github.com/ipfs/testground/sdk/sync=../sdk/sync",
-		"-replace=github.com/ipfs/testground/sdk/runtime=../sdk/runtime")
-
-	// Write replace directives.
-	cmd := exec.CommandContext(ctx, "go", append([]string{"mod", "edit"}, replaces...)...)
-	cmd.Dir = plandst
-	cmd.Stdout = os.Stdout
-	cmd.Stderr = os.Stderr
-	err = cmd.Run()
-	if err != nil {
-		return nil, fmt.Errorf("unable to add replace directives to go.mod; %w", err)
+	if sdksrc != "" {
+		replaces = append(replaces,
+			"-replace=github.com/ipfs/testground/sdk/sync=../sdk/sync",
```

Comment:
I think we should remove these replace lines.

When the sdk is provided by the client, I can't say for certain that the SDK is `github.com/ipfs/testground/...` Even if it was, it is possible that whatever local edits the user has made might not be satisfied with just these replace lines.

What I am thinking is that the SDK should be renamed to `replaces` or some such. This way, the user can add replacements in their own go mod to replace the SDK or whatever other code they want to add replacements. The contract should be that this directory will sit next to the plans directory and your replace lines will be used as you have written them. What do you think of this?

Reply:
> When the sdk is provided by the client, I can't say for certain that the SDK is github.com/ipfs/testground/...

Not sure I follow. The SDK always has a fixed import path. That's how test plans import it (`github.com/ipfs/testground/sdk/{runtime,sync}` pre-refactor, `github.com/ipfs/testground/sdk` after #858).

And this is a Go builder, so it knows how to apply that override in Go. A JS builder will know the import path of its SDK and will use npm/package.json fiddling appropriately.

Reply:
> What I am thinking is that the SDK should be renamed to replaces or some such. This way, the user can add replacements in their own go mod to replace the SDK or whatever other code they want to add replacements. The contract should be that this directory will sit next to the plans directory and in the same way it is on the client's local TESTGROUND_HOME. This way, the user can write their own replace entries if they want to provide customizations to the SDK or any other component, and their go.mod will make sense on the client side and when the daemon does the builds.

The SDK really has a special standing here, because you code a test plan against an SDK.

Overriding arbitrary dependencies is useful for when you run a test plan against version X or Y of an upstream dependency, e.g. go-libp2p v0.2.2 vs v0.3.0.

To make that functionality really useful, we need to allow for overriding dependencies **at the group level**, so you can compose hybrid test runs with groups A and B, where the former hits one version, and the latter hits another.

We already have that functionality: users can specify upstream dependency overrides via the `groups.build.dependencies` composition field. See it in action here with the DHT test plan: https://github.com/ipfs/testground/blob/master/compositions/all-cypress-k8s.toml#L126-L132

The only shortcoming is that the user cannot ship arbitrary local code for that replace, it needs to live under a commit in the origin repo of that dependency.

That _might_ be useful; I'm just now sure how much. Let's stash the idea for now, and not intertwine it with SDK replaces, which are plan-wide, not group-scoped.